### ISSUE TITLE
Skip redundant-returns-doc if function always raises

### ISSUE THREAD

**Issue by jack1142** [PyCQA/pylint#7498]
### Current problem

Suppose you had a `Klass` class:
```py
# pylint: disable=missing-docstring

class Klass:
    def some_method(self) -> str:
        return "Klass"

    def optional_method(self) -> str:
        """
        Optional method which gives [some text].

        Returns
        -------
        str
            [Some text].

        Raises
        ------
        ValueError
            [Some text] couldn't be generated.
        """
        raise ValueError("Default implementation can't generate [some text].")
```
which is meant to be subclassed by your users and has a method that can be *optionally* overridden (like `optional_method()` above).
When users are overriding `optional_method` in their subclass of `Klass`, they are expected to return some value from it. To make it clear how the method is supposed to work, I would want to be able to document the expected return type along with the expected behavior of the method that users should implement if they're overriding the method.
Currently, when using `pylint.extensions.docparams` extension, I get `redundant-returns-doc` warning on such methods:
```
❯ pylint repro.py --load-plugins=pylint.extensions.docparams
************* Module repro
repro.py:7:4: W9008: Redundant returns documentation (redundant-returns-doc)

------------------------------------------------------------------
Your code has been rated at 8.00/10 (previous run: 8.00/10, +0.00)
```

### Desired solution

I think that pylint should not report `redudant-returns-doc` warning when the implementation always raises.
I imagine that most commonly such methods would be raising `NotImplementedError` *but* since I've seen occurrences in my code base where that's not the case, it would be great if the method always raising some other exception would also cause pylint to not report this warning.
I think that just checking that the method has only a single `raise` statement (aside from the docstring and comments) and nothing else would be enough here and IMO there's no point in doing more advanced inference than that.

**Comment by DanielNoord** [comment]
I don't think this is feasible. We rely on `astroid.nodes.FunctionDef.is_abstract`:
https://github.com/PyCQA/astroid/blob/849d043f3e3cd63881ba919ccba4e6a726947d22/astroid/nodes/scoped_nodes/scoped_nodes.py#L1634

This is a fairly robust and time-tested method which shows whether we should consider something an abstract method and even supports patterns that were common when the `abc` package didn't exist.

We could either:
1) Add `ValueError` or any other exception to the list of "abstract-creating-methods" like `NotImplementedError`, but I don't think that makes sense.
2) or, add an exception in this pylint check for this pattern. However, we have recently had other issues about our "abstract-recognition". I'm not a big fan of having multiple definitions of "abstractness" around our codebase. Relying on the simple method at least ensure that we are using the same definition everywhere.
3) or, use the `any_raise_is_abstract` flag of the `is_abstract` method. However, I think this is way too broad. We can't make a fine-grained choice of exceptions to raise for. Furthermore, I think the same argument applies here. I'm against the use of the flag as it makes our definition of "abstractness" unstable.

Actually I think that the warnings are a good point for anybody using `raise ValueError` to show abstractness to start using another method of doing so.

I would welcome a PR that adds some stuff about this in the `details.rst` of the `redundant-returns-doc` message documentation. It is always good to make this behaviour explicit. But if you're not up for that, which I can totally understand, I think we should close this as a won't fix.

**Comment by jack1142** [comment]
Hmm, I was sure I tested it with NotImplementedError but I guess I didn't because it clearly work when raising that instead :)

Before closing as won't fix (better documentation would be nice so IMO keeping it open would make sense but I *personally* won't work on improving the documentation for this), what do you think about checking MRO for `NotImplementedError`?

**Comment by DanielNoord** [comment]
> what do you think about checking MRO for `NotImplementedError`?

What do you mean by this?

> Before closing as won't fix (better documentation would be nice

I'm not sure myself if we are really lacking here. So without any immediate suggestions I think holding an issue open about a small documentation issue where nobody has an immediate solution just clutters the tracker.

**Comment by jack1142** [comment]
> What do you mean by this?

I mean including subclasses of `NotImplementedError` in the mentioned is abstract check.

**Comment by DanielNoord** [comment]
We could, but that would add additional inference calls to something that's arguably working fine. Is there a reason why subclassing the exception instead of using the normal `NotImplementedError` or just using `abc` is better? I guess you could do:

```python
class PylintException:
    ...

class PylintImplementMe(PylintException, NotImplementedError):
    ...

try:
    ...
except PylintException:
    ...
```

We use a similar pattern to catch all errors caused by `astroid`. However, how many times will a library actually want to raise `PylintImplementMe` or any similar subclass? That just seems like poor design and something the library that's depending on the library should probably catch themselves, which can then be done by just using `NotImplementedError` to begin with.

**Comment by jack1142** [comment]
I guess there isn't much reason. I was thinking of a way to switch from custom exception to NotImplementedError without breakage and replacing the custom exception to subclass from `NotImplementedError` rather than `Exception` seemed like a potential option. But thinking about it more, I might as well remove custom exception and set custom exception's name in the module as an alias for `NotImplementedError` and raise that directly.

**Comment by DanielNoord** [comment]
I'm closing this then! Anybody who finds this issue and has a good suggestion for a documentation improvement is more than welcome to open a PR 😄
